### ISSUE TITLE
`invalid_token`: Authorization errors with Sandbox API

### ISSUE THREAD

**Issue by johnboxall** [SalesforceCommerceCloud/sfcc-ci#88]
Hey folks,

**Thanks so much for this great tool!** 

I'm scripting management of our on-demand instances and having trouble using the `access_token` returned by `sfcc-ci auth:client` to do it.

As I understand it, to allow my script to manage our On Demand instances, I must:

1. Create an API Key in Account Manager, setting appropriate scopes and noting its key and password as [per this article](https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/Sandboxes/APIClientID.html)
2. Install this package 
3. Run something like:
    ```bash
    sfcc-ci auth:client "${AM_API_CLIENT_KEY}" "${AM_API_CLIENT_PASSWORD}"
    sfcc-ci sandboxes:list
   ```

So, following those steps, I first use my API Client to get an `access_token`:

```bash
sfcc-ci auth:client "${AM_API_CLIENT_KEY}" "${AM_API_CLIENT_PASSWORD}"
```
Which succeeds returning:

```json
{
    "access_token": "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa",
    "scope": "tenantFilter profile roles",
    "token_type": "Bearer",
    "expires_in": 1799 
}
```

Then I try to list the sandboxes:

```bash
sfcc-ci sandbox:list
```

Which **fails** with the following error:
```json
{"statusCode": "401",
     "body":
      { "error": "invalid_token",
        "error_description": "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa" }
}
```

🤔 

I should note that if I try the same flow using `auth:login` rather than `auth:client`, things work:

```bash
sfcc-ci auth:login

# Browser opens and successfully auths, CLI acks auth.

sfcc-ci sandbox:list

# CLI successfully lists sandbox instances
```

_This flow wouldn't work however as the intention is to just run a script, not have someone click buttons to auth things :)_

Any ideas?

Thanks so much.

**Comment by appanp-tjs** [comment]
I am getting the following error, I do not think that there is anything wrong with the four input parameters to `sfcc-ci client:auth` command. How can I debug this issue? I have also tried with curl but same error.

> { error_description: 'Resource owner authentication failed',
        error: 'invalid_grant' }

**Comment by appanp-tjs** [comment]
Finally I could find the problem. Looks like you get this error when 2FA is enabled for Sandbox API automation user and there is no paired 2nd factor device, in the **primary organisation of the automation user**. In this case, the auth flow is redirected to the "Connect Salesforce Authentocator" page  when I tried to login using the Swagger Sandbox API app. Once I disabled the "Sandbox API Role" from the ones which require 2FA (for the primary organisation of automation user), I could get the auth bearer token successfully.

**Comment by tobiaslohr** [comment]
@johnboxall can we close this ticket?

**Comment by ashwanik-github** [comment]
> https://gist.github.com/johnboxall/7fd92fb8667314ea4d1628fb5229f162

Yeah. We are also having a similar issue during fetching of access_token on the AM API - https://account.demandware.com/dw/oauth2/access_token 
Response:  401 Unauthorized
`{
    "error_description": "Client authentication failed",
    "error": "invalid_client"
}`

Since we have enabled the Salesforce Authenticator for 2FA/MFA the API calls (below using the curl) fail to fetch the token.
`curl -X POST "https://account.demandware.com/dw/oauth2/access_token" -H 'Content-Type: application/x-www-form-urlencoded'  --data client_id=${SFCC_OAUTH_CLIENT_ID} --data client_secret=${SFCC_OAUTH_CLIENT_SECRET}  --data grant_type=password --data username=${SFCC_OAUTH__USERNAME} --data password=${SFCC_OAUTH_USER_PASSWORD}|jq -r '.access_token'`
Nevertheless, the API call using sfcc-ci triggers MFA notifications but the curl does not. 
Any clue how we can fix this?
Thanks!

**Comment by tobiaslohr** [comment]
`invalid_client` and `invalid_token` are different things, the way how the CLI obtains token through the resource owner password credentials grant is documented through the code as you probably know. There is nothing the CLI explicitly makes use of to trigger MFA. Usually this is specified by the MFA method the user / primary org of the user has set.

If the curl is correct or not, I cannot tell. However, this grant type is subject for deprecation in AM and hence we will remove this flow in the CLI. See also https://github.com/SalesforceCommerceCloud/sfcc-ci/wiki/Resource-Owner-Password-Credentials-Authentication-Failure.

As a consequence, I recommend to not use a service user for your use case anymore, but rather switch to a pure client_credentials grant.

**Comment by esmiralha** [comment]
I get the same error when running client:auth with just the api client id and api client secret.

```
DEBUG=* sfcc-ci client:auth '******' '******'
[DEBUG] Configuration loaded from /Users/****/dw.json
[DEBUG] Authorize via Oauth client_credentials grant
[DEBUG] Doing auth request, payload: {"grant_type":"client_credentials"}
{
  request: {
    debugId: 1,
    uri: 'https://account.demandware.com/dw/oauth2/access_token',
    method: 'POST',
    headers: {
      host: 'account.demandware.com',
      'content-type': 'application/x-www-form-urlencoded',
      authorization: 'Basic *********',
      accept: 'application/json',
      'content-length': 29
    },
    body: 'grant_type=client_credentials'
  }
}
{
  response: {
    debugId: 1,
    headers: {
      'cache-control': 'no-store',
      pragma: 'no-cache',
      'www-authenticate': 'Basic realm="/"',
      'content-type': 'application/json;charset=UTF-8',
      'content-length': '77',
      date: 'Wed, 26 Apr 2023 13:25:04 GMT',
      'referrer-policy': 'strict-origin-when-cross-origin',
      'strict-transport-security': 'max-age=15724800; includeSubDomains',
      'x-content-type-options': 'nosniff',
      'x-frame-options': 'SAMEORIGIN',
      vary: 'Accept-Encoding',
      'set-cookie': [Array],
      server: 'envoy',
      connection: 'close'
    },
    statusCode: 401,
    body: {
      error_description: 'Client authentication failed',
      error: 'invalid_client'
    }
  }
}
Error: Authentication failed: Client authentication failed
```

**Comment by tobiaslohr** [comment]
@esmiralha I think you've solved this already as part of https://sfcc-unofficial.slack.com/archives/CAUFG3SHF/p1682597925151959. The credentials passed in were wrong / got mangled (since not escaped properly), as a result the authentication failed.
